src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,210 @@
+            }
+
+            var token = parameterToken[Methods.ProgressNotificationTokenName].ToObject<string>(); // IProgress<object>>();
+
+            if (string.IsNullOrEmpty(token) || !_activeRequests.TryGetValue(token, out var request))
+            {
+                return;
+            }
+
+            var value = parameterToken[ProgressNotificationValueName];
+
+            try
+            {
+                request.HandlerCancellationToken.ThrowIfCancellationRequested();
+                await request.OnProgressNotifyAsync(value, request.HandlerCancellationToken).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Initial handle request has been cancelled
+                // We can deem this ProgressRequest complete
+                ProgressCompleted(token);
+                return;
+            }
+
+            CompleteAfterDelay(token, request);
+        }
+
+        private void CompleteAfterDelay(string token, ProgressRequest request)
+        {
+            CancellationTokenSource linkedCTS;
+
+            lock (request.RequestLock)
+            {
+                CancelAndDisposeToken(request.TimeoutCancellationTokenSource);
+
+                request.TimeoutCancellationTokenSource = new CancellationTokenSource();
+                linkedCTS = CancellationTokenSource.CreateLinkedTokenSource(
+                    request.TimeoutCancellationTokenSource.Token,
+                    request.HandlerCancellationToken);
+            }
+
+            _ = CompleteAfterDelayAsync(token, request.TimeoutAfterLastNotify, linkedCTS); // Fire and forget
+
+
+            async Task CompleteAfterDelayAsync(string token, TimeSpan delay, CancellationTokenSource cts)
+            {
+                try
+                {
+                    await Task.Delay(delay, cts.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // Task cancelled, new progress notification received.
+                    // Don't allow handler to return
+                    return;
+                }
+                finally
+                {
+                    // Dispose of the Linked Cancellation Token Source
+                    // Note: The component TimeoutCancellationTokenSource & HandlerCancellationToken
+                    //       should not be impacted. Their lifecycle is managed using CancelAndDisposeToken
+                    //       and the platform handler, respectively.
+                    cts.Dispose();
+                }
+
+                ProgressCompleted(token);
+            }
+        }
+
+        private static void CancelAndDisposeToken(CancellationTokenSource cts)
+        {
+            if (cts != null &&
+                cts.Token.CanBeCanceled &&
+                !cts.Token.IsCancellationRequested)
+            {
+                cts.Cancel();
+                cts.Dispose();
```

Comment:
Any harm in always disposing (whether you cancel or not)? If it's already disposed I believe it'll no-op

Reply:
Just tried it out, everything seems to explode. Reverted the change.

This shouldn't be an issue, as the only time we use this function is with the `request.TimeoutCancellationTokenSource`. That token is only cancelled with this function. So when it's cancelled it'll be disposed of. I believe things are exploding because the token is in an un-cancellable state, and we're encountering a race condition of some sort. 

Thread 1: Cancels token

While token is being cancelled, Thread 2 disposes token. This results in a funky edge case where we're cancelling a disposed token.